### PR TITLE
Add devices without restart (fixes #2083)

### LINKED ISSUE #2083: Adding devices to existing folder should not require restart

### PR DESCRIPTION
**Pull request by calmh**
This adds support for adding devices to folders (and new devices in general) without requiring restart.

This shows a blank device receiving a connection from another device, adding it, sharing the default folder with it, and syncing up all without restart:

https://nym.se/t/restartless.mov

(Adding new folders without restart is not yet supported, likewise removing devices from folders etc.)

### CODE REVIEW

**Comment by calmh** [comment]
@st-jenkins run integration tests, please

**Comment by Zillode** [comment]
@st-jenkins test on windows and osx

**Comment by Zillode** [comment]
@st-jenkins try tests on osx again

**Comment by calmh** [comment]
Failure unrelated to this, but I'll look into it...

```
=== RUN TestReset
2015/07/22 08:53:15 Cleaning...
2015/07/22 08:53:15 Creating files...
2015/07/22 08:53:15 Starting instance 1...
2015/07/22 08:53:16 Cleaning...
2015/07/22 08:53:16 Reset invalid folder
2015/07/22 08:53:16 Reset indexes of default folder
2015/07/22 08:53:17 Starting instance 1...
2015/07/22 08:53:19 Creating files...
2015/07/22 08:53:19 Reset DB...
2015/07/22 08:53:20 eventLoop: events: Get http://127.0.0.1:8081/rest/events?since=12: dial tcp 127.0.0.1:8081: connection refused
2015/07/22 08:53:20 Starting instance 1...
2015/07/22 08:53:20 eventLoop: events: Get http://127.0.0.1:8081/rest/events?since=12: dial tcp 127.0.0.1:8081: connection refused
2015/07/22 08:53:20 eventLoop: events: Get http://127.0.0.1:8081/rest/events?since=12: dial tcp 127.0.0.1:8081: connection refused
2015/07/22 08:53:21 eventLoop: events: Get http://127.0.0.1:8081/rest/events?since=12: dial tcp 127.0.0.1:8081: connection refused
2015/07/22 08:53:21 eventLoop: events: Get http://127.0.0.1:8081/rest/events?since=12: dial tcp 127.0.0.1:8081: connection refused
2015/07/22 08:53:21 eventLoop: events: Get http://127.0.0.1:8081/rest/events?since=12: dial tcp 127.0.0.1:8081: connection refused
2015/07/22 08:53:21 eventLoop: events: unexpected EOF
--- FAIL: TestReset (6.65s)
	util.go:527: Post http://127.0.0.1:8081/rest/system/shutdown: dial tcp 127.0.0.1:8081: connection refused
	util.go:527: Post http://127.0.0.1:8081/rest/system/shutdown: dial tcp 127.0.0.1:8081: connection refused
```

File: internal/config/config.go [hunk 1]
```diff
@@ -82,8 +82,6 @@ type FolderConfiguration struct {
 	IgnoreDelete    bool                        `xml:"ignoreDelete" json:"ignoreDelete"`
 
 	Invalid string `xml:"-" json:"invalid"` // Set at runtime when there is an error, not saved
-
-	deviceIDs []protocol.DeviceID
```
Comment (calmh) [inline internal/config/config.go]: We can't cache this as that makes the struct not compare correctly after changing the device list. It's no big deal, we don't call `DeviceIDs()` (that uses this) in any performance sensitive spots.

**Comment by AudriusButkevicius** [comment]
So looks fine (apart from the duplication of map features, perhaps interfaces is the way forward). Why can't we do the same with removal?

**Comment by calmh** [comment]
We can, I just didn't implement that yet. I'll do that, and adding/removing folders next.

**Comment by AudriusButkevicius** [comment]
I guess my only concern in the code duplication as I suggested, after that looks good to merge.

**Comment by calmh** [comment]
You mean the various `map*`-functions? It doesn't really bother me as they are small and obvious and kinda need to exist as they take and return different types and Go doesn't have generics (*cough cough*)... I guess an alternative would be to have them on the slice types, i.e. `type DeviceIDSlice []DeviceID` and `func (s DeviceIDSlice) ToMap() map[DeviceID]struct{}` or something, but meh...

**Pull request merged**